### PR TITLE
entropy target algorithm

### PR DESCRIPTION
**Pull request by emailweixu**
This implements the method described in section 5 of "Soft Actor-Critic Algorithms and Applications".
It automatically adjust weight for entropy loss term so that entropy is not smaller than the given target.

### CODE REVIEW

File: alf/algorithms/entropy_target_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+import gin
+import numpy as np
+import tensorflow as tf
+
+from tf_agents.agents.tf_agent import LossInfo
+from tf_agents.utils import common as tfa_common
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep
+
+from alf.utils import common, dist_utils
```
Comment (witwolf) [inline alf/algorithms/entropy_target_algorithm.py]: `dist_uitls`  not added in this commit

File: alf/algorithms/entropy_target_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            optimizer=optimizer,
+            debug_summaries=debug_summaries,
+            name="EntropyTargetAlgorithm")
+
+        self._log_alpha = log_alpha
+        self._action_spec = action_spec
+
+        def _calc_default_target_entropy(spec):
+            dims = np.product(spec.shape.as_list())
+            if spec.dtype.is_floating:
```
Comment (witwolf) [inline alf/algorithms/entropy_target_algorithm.py]: it's more clear to use `tensor_spec.is_continuous`

File: alf/algorithms/entropy_target_algorithm.py (hunk 1 above)
Comment (witwolf) [reply alf/algorithms/entropy_target_algorithm.py]: sorry, ignore comment above  `dist_utils` already exists in mater

File: alf/algorithms/entropy_target_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        return AlgorithmStep(
+            outputs=(),
+            state=(),
+            info=LossInfo(
+                loss,
+                extra=EntropyTargetLossInfo(
+                    alpha_loss=alpha_loss, entropy_loss=entropy_loss)))
+
+    def calc_loss(self, training_info):
+        if self._debug_summaries and common.should_record_summaries():
```
Comment (witwolf) [inline alf/algorithms/entropy_target_algorithm.py]: Is  condition `common.should_record_summaries()`  needed here as it already run under summary context with `run_under_record_context` ?

File: alf/algorithms/entropy_target_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            AlgorithmStep. `info` field is LossInfo, other fields are empty.
+        """
+        entropy, entropy_for_gradient = dist_utils.entropy_with_fallback(
+            distribution, self._action_spec)
+        alpha_loss = self._log_alpha * tf.stop_gradient(entropy -
+                                                        self._target_entropy)
+        alpha = tf.stop_gradient(tf.exp(self._log_alpha))
+        loss = alpha_loss
+        entropy_loss = -entropy
+        loss -= alpha * entropy_for_gradient
```
Comment (witwolf) [inline alf/algorithms/entropy_target_algorithm.py]: it's better to add comment for this joint loss and  add `target_entropy_loss` for  `EntropyTargetLossInfo`

File: alf/algorithms/entropy_target_algorithm.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/entropy_target_algorithm.py]: updated

File: alf/algorithms/entropy_target_algorithm.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/entropy_target_algorithm.py]: It shouldn't affect the summary. But it may save a little computation for preparing the value to be summarized.

File: alf/algorithms/entropy_target_algorithm.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/entropy_target_algorithm.py]: added
